pkg/controllers/user/globaldns/user_cluster_globaldns_controller.go

```diff
@@ -56,67 +77,130 @@ func (g *UserGlobalDNSController) sync(key string, obj *v3.GlobalDNS) (runtime.O
+			allIngresses = append(allIngresses, ingresses...)
+		}
 	}
+	//gather endpoints
+	return g.fetchGlobalDNSEndpointsForIngresses(allIngresses, obj)
+}
+
+func (g *UserGlobalDNSController) fetchGlobalDNSEndpointsForIngresses(ingresses []*v1beta1.Ingress, obj *v3.GlobalDNS) ([]string, error) {
 	if len(ingresses) == 0 {
 		return nil, nil
 	}
 
+	var allEndpoints []string
+	//gather endpoints from all ingresses
 	for _, ing := range ingresses {
 		if gdns, ok := ing.Annotations[annotationGlobalDNS]; ok {
 			// check if the globalDNS in annotation is same as the FQDN set on the GlobalDNS
 			if gdns != obj.Spec.FQDN {
 				continue
 			}
-			//update endpoints on GlobalDNS status field
+			//gather endpoints from the ingress
 			ingressEndpoints := gatherIngressEndpoints(ing.Status.LoadBalancer.Ingress)
-			updatedGDNS, err := g.updateGlobalDNSEndpoints(obj, ingressEndpoints)
-			if err != nil {
-				return updatedGDNS, err
-			}
+			allEndpoints = append(allEndpoints, ingressEndpoints...)
 		}
 	}
-	return nil, nil
+
+	return allEndpoints, nil
 }
 
-func (g *UserGlobalDNSController) updateGlobalDNSEndpoints(globalDNS *v3.GlobalDNS, ingressEndpoints []string) (*v3.GlobalDNS, error) {
-	globalDNS = prepareGlobalDNSForUpdate(globalDNS, ingressEndpoints)
-	updated, err := g.globalDNSs.Update(globalDNS)
+func (g *UserGlobalDNSController) refreshGlobalDNSEndpoints(globalDNS *v3.GlobalDNS, ingressEndpointsForCluster []string) (*v3.GlobalDNS, error) {
+
+	mapIngressEndpoints := make(map[string]bool)
+	for _, ep := range ingressEndpointsForCluster {
+		mapIngressEndpoints[ep] = true
+	}
+	//find endpoints removed by looking at clusterEndpoints of current cluster
+	endpointsToRemove := []string{}
+
+	if len(globalDNS.Status.ClusterEndpoints) == 0 {
+		globalDNS.Status.ClusterEndpoints = make(map[string][]string)
+	}
+
+	for _, ep := range globalDNS.Status.ClusterEndpoints[g.clusterName] {
+		if !mapIngressEndpoints[ep] {
+			endpointsToRemove = append(endpointsToRemove, ep)
+		}
+	}
+
+	globalDNSToUpdate := globalDNS.DeepCopy()
+	globalDNSToUpdate.Status.ClusterEndpoints[g.clusterName] = ingressEndpointsForCluster
+	//update endpoints on GlobalDNS status field
+	updatedGDNS, err := g.updateGlobalDNSEndpoints(globalDNSToUpdate, ingressEndpointsForCluster, endpointsToRemove)
 	if err != nil {
-		return updated, fmt.Errorf("UserGlobalDNSController: Failed to update GlobalDNS endpoints with error %v", err)
+		return updatedGDNS, err
 	}
-	return nil, nil
+
+	return updatedGDNS, nil
 }
 
-func getMultiClusterAppName(multiClusterAppName string) (string, error) {
-	split := strings.SplitN(multiClusterAppName, ":", 2)
-	if len(split) != 2 {
-		return "", fmt.Errorf("error in splitting multiclusterapp ID %v", multiClusterAppName)
+func (g *UserGlobalDNSController) updateGlobalDNSEndpoints(globalDNS *v3.GlobalDNS, ingressEndpoints []string, endpointsToRemove []string) (*v3.GlobalDNS, error) {
+	prepareGlobalDNSForUpdate(globalDNS, ingressEndpoints, g.clusterName)
```

Comment:
I wonder if it would be more clean if `globalDns.Endpoints` are reconciled separately. So we don't have to figure out the delta (toAdd/toRemove). Instead, it should just be a aggregation of all the clusterEndpoints.

Reply:
since we already know `endpointsToRemove`, can we just make it one function and calculate final endpoints just once before update?

Reply:
have refactored this code, we no longer need these methods